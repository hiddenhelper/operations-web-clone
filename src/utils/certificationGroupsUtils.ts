import { CertificationModel } from 'modules/models';
import { IProjectCertification } from 'modules/models/certification';

export const sanitizeCertificationGroups = (certificationGroups: CertificationModel.ICertificationGroup[]): CertificationModel.ICertificationGroup[] => {
  return certificationGroups.map((certificationGroup: CertificationModel.ICertificationGroup) => ({
    ...certificationGroup,
    certifications: certificationGroup.certifications.map((certification: IProjectCertification) => ({
      ...certification,
      alias: certification.alias?.trim() || null,
    })),
  }));
};
