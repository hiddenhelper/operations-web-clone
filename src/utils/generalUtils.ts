import moment, { Moment } from 'moment';
import createNumberMask from 'text-mask-addons/dist/createNumberMask';

import { GeneralModel, ResourceModel, StatisticsModel, UserModel } from '../modules/models';
import { stateMap } from '../constants/territory';
import { formatIncompletePhoneNumber } from 'libphonenumber-js';

export const EMAIL_REGEXP = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;

const PHONE_MASK = '(...) ...-....';
export const PHONE_MASK_LIST = { mx: PHONE_MASK, pr: PHONE_MASK };

const CANADA_TAX_ID_REGEX = /^((\d{2})([-])(\d{7})|(\d{3})([-])(\d{2})([-])(\d{4}))$/;
const MEXICO_TAX_ID_REGEX = /^[A-Za-z]{3,4}[-]{0,1}[0-9]{6}[-]{0,1}[0-9A-Za-z]{3}$/;
const USA_TAX_ID_REGEX = /^(\d{2})([-])(\d{7})$/;

export const TAX_ID_REGEXP_LIST = [CANADA_TAX_ID_REGEX, MEXICO_TAX_ID_REGEX, USA_TAX_ID_REGEX];

export const TAG_ID_MASK = Array(12).fill(/[0-9A-Fa-f]/);

export const getFontBaseValue = (defaultFontSize = 16): number => {
  return defaultFontSize;
};

export const toREM = (value: number): string => {
  const baseValue = getFontBaseValue();
  return `${value * (1 / baseValue)}rem`;
};

export const hexToRgb = (hex: string): string | null => {
  const hexRegex = /^#?([a-f\d])([a-f\d])([a-f\d])$/i;
  const rgb = hex.replace(hexRegex, (m, r, g, b) => `${r}${r}${g}${g}${b}${b}`);

  const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(rgb);
  const toBase16 = (value: string) => parseInt(value, 16);

  return result ? `${toBase16(result[1])}, ${toBase16(result[2])}, ${toBase16(result[3])}` : null;
};

export const formatBytes = (bytes: number, decimals: number = 2): string => {
  if (bytes === 0) return '0 B';

  const k = 1000;
  const dm = decimals < 0 ? 0 : decimals;
  const sizes = ['B', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'];

  const i = Math.floor(Math.log(bytes) / Math.log(k));

  return `${parseFloat((bytes / Math.pow(k, i)).toFixed(dm))} ${sizes[i]}`;
};

export const isObject = entity => entity && typeof entity === 'object' && entity.constructor === Object;

export const isArray = value => Array.isArray(value);

export const isUndefined = value => value === undefined;

export const isNull = value => value === null;

export const isNumber = value => typeof value === 'number';

export const isString = value => typeof value === 'string';

export const isEmpty = value => {
  if (isObject(value)) {
    return Object.keys(value).length === 0;
  }
  if (isUndefined(value) || isNull(value)) {
    return true;
  }
  if (isArray(value)) {
    return value.length === 0;
  }
  if (isNumber(value)) {
    return Number.isNaN(value);
  }
  if (isString(value)) {
    return value.trim() === '';
  }

  return value === '';
};

export const isUUID = id => /^[0-9a-f]{8}-[0-9a-f]{4}-[0-5][0-9a-f]{3}-[089ab][0-9a-f]{3}-[0-9a-f]{12}$/i.test(id);

export const parseNumber = (num: number) => parseInt(String(num).replace(/[,]/g, ''), 10);
export const parseFloatNumber = (num: number) => parseFloat(String(num).replace(/[,]/g, ''));

export const roundNumber = (num: number) => Math.round(num * 100) / 100;

export const injectScript = ({ src, scriptId }: { src: string; scriptId: string }) => {
  if (document.getElementById(scriptId)) return Promise.resolve(document.getElementById(scriptId));
  return new Promise(function(resolve, reject) {
    const script = document.createElement('script');
    script.src = src;
    script.id = scriptId;
    script.type = 'text/javascript';
    script.onload = /* istanbul ignore next */ () => resolve(script);
    script.onerror = /* istanbul ignore next */ () => reject(new Error(`Script load error for ${src}`));
    document.body.append(script);
  });
};

export const disableFocus = () => {
  document.body.classList.add('using-mouse');
  return true;
};

export const enableFocus = () => {
  document.body.classList.remove('using-mouse');
  return true;
};

export const enableFocusOnTab = event => {
  return event.keyCode === 9 ? enableFocus() : false;
};

export const getPrevObjectItem = (obj, currentKey) => {
  const keys = Object.keys(obj);
  const index = keys.indexOf(currentKey);
  const prevKey = index - 1;
  return prevKey < 0 ? obj[currentKey] : obj[keys[prevKey]];
};

export const getNextObjectItem = (obj, currentKey) => {
  const keys = Object.keys(obj);
  const index = keys.indexOf(currentKey);
  const nextKey = index + 1;
  return nextKey >= keys.length ? obj[currentKey] : obj[keys[nextKey]];
};

export const getCompletedStepFields = (stepMap: GeneralModel.IStepMap, currentEntity): GeneralModel.IStepCompletedMap => {
  return Object.entries(stepMap).reduce(
    (accumulator, [key, item]) => ({
      ...accumulator,
      [item.key]: {
        title: item.title,
        order: item.order,
        required: item.fields.reduce((acc, f) => {
          if (f.requiredFunc) return f.requiredFunc({ field: f, accumulator: acc, value: currentEntity[f.name] });
          if (f.fields && Array.isArray(currentEntity[f.name])) {
            const multipleBy = currentEntity[f.name].length > 1 ? currentEntity[f.name].length : 1;
            return acc + f.fields.filter(ff => ff.required).length * multipleBy;
          } else if (f.fields) {
            return acc + f.fields.filter(ff => ff.required).length;
          }
          return f.required ? acc + 1 : acc + 0;
        }, 0),
        completed: item.fields.reduce((acc, f) => {
          if (f.completedFunc) return f.completedFunc({ field: f, accumulator: acc, value: currentEntity[f.name] });
          if (f.fields && Array.isArray(currentEntity[f.name])) {
            currentEntity[f.name].forEach(currentItem => {
              acc = f.fields.map(field => field.required && (field.computePositive || !isEmpty(currentItem[field.name]))).filter(Boolean).length + acc;
            });
            return acc;
          }
          if (f.fields && isObject(currentEntity[f.name])) {
            return (
              f.fields
                .map(field => field.required && (field.computePositive || !isEmpty(currentEntity[f.name] && currentEntity[f.name][field.name])))
                .filter(Boolean).length + acc
            );
          }
          return f.required && (f.computePositive || !isEmpty(currentEntity && currentEntity[f.name])) ? acc + 1 : acc + 0;
        }, 0),
      },
    }),
    {}
  );
};

export const isTempId = (_id: string): boolean => /^temp-/.test(_id);

export const generateTempId = (() => {
  let counter = 0;
  return () => `temp-${++counter}`;
})();

export const parseUrlSearch = <T = { [key: string]: string }>(search: string, initialValue: T = {} as any): T =>
  search
    .replace('?', '')
    .split('&')
    .map(s => s.split('='))
    .reduce(
      (total, [key, value]) => ({
        ...total,
        [key]: value,
      }),
      initialValue
    );

export const safeParse = <T>(item: any, fallback?: any): T => {
  try {
    return JSON.parse(item);
  } catch {
    return fallback !== undefined ? fallback : item;
  }
};

export const formatFreedomBadgeCode = (code: string) => code.replace(/(\w{1,2})(\w{2})(\w{2})/, '$1 $2 $3');
export const formatBadgeCode = (code: string) => (code?.length === 6 ? formatFreedomBadgeCode(code) : code);

export const getDrawerButton = (
  status: ResourceModel.Status,
  id: string,
  type: ResourceModel.Type.CLIENT | ResourceModel.Type.PROJECT,
  version?: string
): { buttonText: string; linkTo: string } => {
  const to = type === ResourceModel.Type.CLIENT ? 'clients' : 'projects';
  const drawerButton = {
    [ResourceModel.Status.DRAFT]: {
      buttonText: 'Edit',
      linkTo: version === 'new' ? `/${to}/wizard-new/${id}` : `/${to}/wizard/${id}`,
    },
    [ResourceModel.Status.ACTIVE]: {
      buttonText: `${type} Detail`,
      linkTo: version === 'new' ? `/${to}/detail-new/${id}` : `/${to}/detail/${id}`,
    },
    [ResourceModel.Status.PENDING_APPROVAL]: {
      buttonText: 'Review',
      linkTo: version === 'new' ? `/${to}/wizard-new/${id}/review` : `/${to}/wizard/${id}/review`,
    },
    [ResourceModel.Status.ARCHIVED]: {
      buttonText: `${type} Detail`,
      linkTo: version === 'new' ? `/${to}/detail-new/${id}` : `/${to}/detail/${id}`,
    },
  };
  return drawerButton[status] || drawerButton[ResourceModel.Status.ACTIVE];
};

export const deleteObjectItem = <T>(obj: T, id) => Object.fromEntries(Object.entries(obj).filter(([itemId, value]) => itemId !== id));

export const numberMask = createNumberMask({
  prefix: '',
  suffix: '',
  allowDecimal: true,
});

export const moneyMask = createNumberMask({
  prefix: '$ ',
  suffix: '',
  allowDecimal: true,
});

export const percentageMask = createNumberMask({
  prefix: '',
  suffix: '%',
  allowDecimal: true,
});

export const plainNumberMask = createNumberMask({
  prefix: '',
  suffix: '',
  allowDecimal: true,
  includeThousandsSeparator: false,
});

export const sanitizeNumber = value =>
  parseFloat(
    String(value)
      .replace(/[$,%]/g, '')
      .replace(' ', '')
  );

export const formatNumberWithCommas = num => (!isEmpty(num) ? num.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',') : '');

export const getDefaultValue = (value: any, defaultValue: any = '-') => (isEmpty(value) ? defaultValue : value);

export const getConditionalDefaultValue = (condition: any, value: any, defaultValue: any = '-') =>
  !!condition ? getDefaultValue(value, defaultValue) : defaultValue;

export const getDashIfNull = (value: any, defaultValue: any) => {
  if (isNull(value)) {
    return '-';
  }
  return defaultValue;
};

export const getFormattedDate = (date: number | string, format: GeneralModel.DateFormat) =>
  date ? moment(date).format(GeneralModel.dateFormatMap[format]) : null;

export const formatIdNumber = (id: string) => `${id.slice(0, 3)}-${id.slice(3, 5)}-${id.slice(5)}`;

export const getFormattedOptionList = (item: GeneralModel.INamedEntity): GeneralModel.ISelectOption => ({ value: item.id, label: item.name });

export const getOptionListFromMap = map => Object.entries(map).map(([key, value]) => ({ value: key, label: value as any }));

export const getOptionListFromNamedEntities = list => list.map(option => getFormattedOptionList(option));

export const getListDefaultValue = (list: { [key: string]: string[] } = {}, key: string, defaultValue: string = '') => {
  return list[key] ? list[key][0] : defaultValue;
};

export const getHideOverflowClass = (list: any[], ref: any, className: string) =>
  list.length < 5 && ref.current && ref.current.offsetHeight >= 395 ? className : '';

export const getListWithCommas = (listItems: GeneralModel.INamedEntity[]) => {
  const formattedList = listItems.map((item, index) => item.name).join(', ');
  return `${formattedList}${getConditionalDefaultValue(formattedList.length, '.', '-')}`;
};

export const getImageUrl = (file: Partial<File>) => (file ? URL.createObjectURL(file) : null);

export const getFileNames = files =>
  Object.entries(files).reduce(
    (totalFileNames, [currentFileKey, currentFileValue]) => ({
      ...totalFileNames,
      [currentFileKey]: (currentFileValue as any).file.name,
    }),
    {}
  );

export const getBooleanDefaultValue = (currentValue: boolean, truthyValue: string, falsyValue: string, defaultValue = '-') =>
  isEmpty(currentValue) ? defaultValue : currentValue ? truthyValue : falsyValue;

export const sortByCreatedDate = (a, b) => (b.createdAt || '').localeCompare(a.createdAt || '');
export const sortByOrder = (a, b) => a.order - b.order;

export const formatPhoneNumber = (number: string) => {
  return !isEmpty(number) ? formatIncompletePhoneNumber(number) : '-';
};

export const sanitizePhoneNumber = (number: string) =>
  String(number)
    .replace(/[()-]/g, '')
    .replace(/ /g, '');

export const getDateFromMomentDateTime = (date: Moment, time: Moment) =>
  moment(date).set({
    hour: time.get('hour'),
    minute: time.get('minute'),
    second: 0,
    millisecond: 0,
  });

export const getTimeFromNow = (dateTime: string) => {
  return moment(new Date(moment.utc(dateTime).format())).fromNow();
};

export const parseQuery = (query): string =>
  Object.keys(query)
    .filter(key => query[key] !== undefined)
    .map(key =>
      Array.isArray(query[key])
        ? query[key].map(item => `${key}=${item}`).join('&')
        : `${key}=${query[key] && typeof query[key] === 'object' ? JSON.stringify(query[key]) : query[key]}`
    )
    .join('&');

export const getQueryParamsFromString = queryString => {
  try {
    const query = {};
    const pairs = (queryString[0] === '?' ? queryString.substr(1) : queryString).split('&');
    for (var i = 0; i < pairs.length; i++) {
      var pair = pairs[i].split('=');
      query[decodeURIComponent(pair[0])] = JSON.parse(decodeURIComponent(pair[1] || ''));
    }
    return query;
  } catch {
    return '';
  }
};

export const sanitizePaginationQuery = ({ page = 1, limit = 30, ...rest }: any) => ({
  pageNumber: page,
  pageSize: limit,
  ...rest,
});

export const getFormattedDecimalNumber = (value: number) => (isEmpty(value) ? null : value.toFixed(2));

export const getVisitorBadgeText = badgeNumber => `VISITOR #${getConditionalDefaultValue(badgeNumber, String(badgeNumber).padStart(4, '0'), '')}`;

export const noop = () => {
  /**/
};

export const getLineChartTicksByData = tick => {
  if (tick.ticks.length > 8 && tick.ticks.length < 20) {
    return tick.index % 3 === 0 ? 'block' : 'none';
  }
  if (tick.ticks.length > 20) {
    return tick.index % 6 === 0 ? 'block' : 'none';
  }
  return 'block';
};

export const getBigLineChartTicksByData = tick => {
  if (tick.ticks.length > 15 && tick.ticks.length < 20) {
    return tick.index % 3 === 0 ? 'block' : 'none';
  }
  if (tick.ticks.length > 20) {
    return tick.index % 5 === 0 ? 'block' : 'none';
  }
  return 'block';
};

export const sanitizeAccountData = (data: UserModel.IAccount) => ({
  ...data,
  preferredContactMethod: isEmpty(data.preferredContactMethod) ? null : Number(data.preferredContactMethod),
  mobilePhoneNumber: isEmpty(data.mobilePhoneNumber) ? null : sanitizePhoneNumber(data.mobilePhoneNumber),
  officePhoneNumber: isEmpty(data.officePhoneNumber) ? null : sanitizePhoneNumber(data.officePhoneNumber),
});

export const toggleArray = (list, item) => {
  const currentList = [...list];
  const index = currentList.indexOf(item);
  if (index !== -1) {
    currentList.splice(index, 1);
    return currentList;
  }
  currentList.push(item);
  return currentList;
};

export const getPieData = (data: StatisticsModel.IPieChartData[], dataMap: any, colorMap: { [key: number]: string }) => {
  const pieData = [...data];
  const notDefinedIndex = pieData.findIndex(entry => entry.segment === -1);
  let notDefinedElement = null;
  if (notDefinedIndex > -1) {
    notDefinedElement = pieData[notDefinedIndex];
    pieData.splice(notDefinedIndex, 1);
  }
  let result = pieData.map((entry, index) => ({
    ...entry,
    x: dataMap[entry.segment],
    y: entry.total,
    color: colorMap[index],
  }));
  if (notDefinedIndex > -1) {
    result = [
      ...result.slice(0, notDefinedIndex),
      {
        ...notDefinedElement,
        x: dataMap[notDefinedElement.segment],
        y: notDefinedElement.total,
        color: '#AAAAAA',
      },
      ...result.slice(notDefinedIndex),
    ];
  }
  return result;
};

export const getLocationList = () => [{ id: '', name: 'All Locations' }, ...Object.entries(stateMap).map(([key, value]) => ({ id: key, name: value }))];
