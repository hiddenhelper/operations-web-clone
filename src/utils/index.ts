export * from './generalUtils';
export * from './useValidator';
