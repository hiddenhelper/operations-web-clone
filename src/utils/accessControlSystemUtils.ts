import { AccessControlSystemModel } from '../modules/models';
import { getConditionalDefaultValue, isEmpty } from './generalUtils';

export const sanitizeReader = (reader: AccessControlSystemModel.IReader): any => {
  let modifiedReader = { ...reader };
  if (reader?.serialNumber) {
    modifiedReader.serialNumber = reader.serialNumber.toLowerCase().trim();
  }
  return {
    ...modifiedReader,
    directionType: getConditionalDefaultValue(!isEmpty(reader?.directionType), Number(reader?.directionType), null),
    sshConnectionPort: getConditionalDefaultValue(reader?.sshConnectionPort, Number(reader?.sshConnectionPort), 0),
    telnetConnectionPort: getConditionalDefaultValue(reader?.telnetConnectionPort, Number(reader?.telnetConnectionPort), 0),
    httpConnectionPort: getConditionalDefaultValue(reader?.httpConnectionPort, Number(reader?.httpConnectionPort), 0),
    tcpConnectionPort: getConditionalDefaultValue(reader?.tcpConnectionPort, Number(reader?.tcpConnectionPort), 0),
  };
};

export const sanitizeACS = (acs: AccessControlSystemModel.IAccessControlSystem) => {
  let modifiedACS = { ...acs };
  if (acs?.reader1SerialNumber) {
    modifiedACS.reader1SerialNumber = acs.reader1SerialNumber.toLowerCase().trim();
  }
  if (acs?.reader2SerialNumber) {
    modifiedACS.reader2SerialNumber = acs.reader2SerialNumber.toLowerCase().trim();
  }
  return {
    ...modifiedACS,
    serialNumber: acs.serialNumber?.trim().toLowerCase() || null,
    reader1: sanitizeReader(acs.reader1),
    reader2: acs.reader2 ? sanitizeReader(acs.reader2) : null,
  };
};

export const sanitizeAssignAcs = (acs: AccessControlSystemModel.IProjectAccessControlSystem) => ({
  ...acs,
  hasReverseInstallation: getConditionalDefaultValue(acs?.hasReverseInstallation, acs?.hasReverseInstallation, false),
  reader1: sanitizeReader(acs.reader1),
  reader2: getConditionalDefaultValue(acs.reader2, sanitizeReader(acs.reader2), null),
});
