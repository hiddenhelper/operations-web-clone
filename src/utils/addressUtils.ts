import { AddressModel } from '../modules/models';
import { isEmpty, getDefaultValue, getConditionalDefaultValue } from './generalUtils';
import { validateObject } from './formUtils';
import { FormRules, CountryCode } from '../constants';

export const canadaRegex = new RegExp(/^[a-zA-Z]{1}[0-9]{1}[a-zA-Z]{1}-[0-9]{1}[a-zA-Z]{1}[0-9]{1}$/);
export const zipRegex = new RegExp(/^[0-9]{5}-[0-9]{4}$|^[0-9]{5}$/);

export const formatZipCode = (postalCode: string) => {
  return postalCode?.trim()?.replace(' ', '-');
};

export const sanitizeZipCode = (zipCode: string) => {
  return getConditionalDefaultValue(zipCode?.length === 6, zipCode?.replace('-', '').replace(' ', ''), zipCode);
};

export const sanitizeAddress = (address: AddressModel.IAddress) => ({
  ...address,
  line1: !isEmpty(address.line1) ? address.line1 : null,
  line2: !isEmpty(address.line2) ? address.line2 : null,
  stateCode: !isEmpty(address.stateCode) ? address.stateCode : null,
  stateName: !isEmpty(address.stateName) ? address.stateName : null,
  countryId: getConditionalDefaultValue(address?.countryId, address?.countryId, getDefaultValue(address?.country?.id, null)),
  city: !isEmpty(address.city) ? address.city : null,
  zipCode: !isEmpty(address.zipCode) ? sanitizeZipCode(address.zipCode) : null,
});

export const validateAddress = ({ value }: { value: AddressModel.IAddress }) => {
  return validateObject<AddressModel.IAddress>(value, FormRules.address.addressRules);
};

export const validateRequiredAddress = ({ value }: { value: AddressModel.IAddress }) => {
  return validateObject<AddressModel.IAddress>(value, FormRules.address.requiredAddressRules);
};

export const isNYState = (stateCode: string) => stateCode === 'NY';

export const isNYCity = (city: string) => city === 'New York';

export const getLocationString = (city: string, stateName: string) =>
  isEmpty(city) && isEmpty(stateName) ? '-' : `${city ?? ''}${city && stateName ? ', ' : ''}${stateName ? stateName : ''}`;

export const getAddressString = (address: AddressModel.IAddress) =>
  address ? `${address.line1 ? `${address.line1}, ` : ''}${getLocationString(address.city, address?.stateName)}` : '-';

export const getZipCodeMaskByCountry = (countryCode: CountryCode = CountryCode.UNITED_STATES, zipCode: string): { placeholder: string; mask: any[] } => {
  const zipCodeMap = {
    [CountryCode.UNITED_STATES]: getConditionalDefaultValue(
      zipCode?.length > 5,
      AddressModel.zipCodeFormatMap[AddressModel.ZipCodeFormatType.FIVE_DIGITS_PLUS_FOUR],
      AddressModel.zipCodeFormatMap[AddressModel.ZipCodeFormatType.FIVE_DIGITS]
    ),
    [CountryCode.PUERTO_RICO]: getConditionalDefaultValue(
      zipCode?.length > 5,
      AddressModel.zipCodeFormatMap[AddressModel.ZipCodeFormatType.FIVE_DIGITS_PLUS_FOUR],
      AddressModel.zipCodeFormatMap[AddressModel.ZipCodeFormatType.FIVE_DIGITS]
    ),
    [CountryCode.MEXICO]: AddressModel.zipCodeFormatMap[AddressModel.ZipCodeFormatType.FIVE_DIGITS],
    [CountryCode.CANADA]: AddressModel.zipCodeFormatMap[AddressModel.ZipCodeFormatType.CANADA_FORMAT],
  };
  return zipCodeMap[countryCode];
};
