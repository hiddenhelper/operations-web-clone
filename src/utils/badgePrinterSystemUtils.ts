import { BadgePrintingSystemModel } from '../modules/models';
import { isEmpty, sanitizeNumber } from './generalUtils';

export const sanitizeDevice = device => ({
  ...device,
  price: !isEmpty(device.price) ? sanitizeNumber(device.price) : null,
});

export const sanitizePrinter = (printer: BadgePrintingSystemModel.IPrinter) => ({
  ...sanitizeDevice(printer),
  type: parseInt(printer.type as any, 10),
});

export const sanitize = (badgePrinterSystem: BadgePrintingSystemModel.IBadgePrintingSystem) => ({
  ...badgePrinterSystem,
  printer: sanitizePrinter(badgePrinterSystem.printer),
  laptop: sanitizeDevice(badgePrinterSystem.laptop),
  scanner: sanitizeDevice(badgePrinterSystem.scanner),
});
