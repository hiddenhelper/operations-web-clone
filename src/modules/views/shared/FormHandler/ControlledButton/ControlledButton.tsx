import React, { memo, useCallback } from 'react';
import { disableFocus, enableFocusOnTab } from '../../../../../utils/generalUtils';

export interface IControlledInput {
  children: React.ReactNode;
  styleClass?: any;
}

const ControlledButton = ({ children, styleClass = '' }: IControlledInput) => {
  const disableFocusHandler = useCallback(() => disableFocus(), []);
  const enableFocusHandler = useCallback(event => enableFocusOnTab(event), []);

  return (
    <span data-testid="controlled-button" tabIndex={-1} className={styleClass} onMouseDown={disableFocusHandler} onKeyDown={enableFocusHandler}>
      {children}
    </span>
  );
};

export default memo(ControlledButton);
