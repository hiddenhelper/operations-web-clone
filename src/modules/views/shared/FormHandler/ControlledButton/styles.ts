import { createStyles, makeStyles } from '@material-ui/core/styles';
import { STYLE } from '../../../../../constants';
import { toREM } from '../../../../../utils/generalUtils';

export const useStyles = makeStyles(theme =>
  createStyles({
    saveButton: {
      width: `${toREM(163)}`,
      height: `${toREM(36)}`,
      fontSize: `${toREM(STYLE.FONT.SIZE.PARAGRAPH)}`,
      lineHeight: `${toREM(17)}`,
      textTransform: 'none',
      boxShadow: 'none',
      fontWeight: 600,
      outline: 'none',
      marginLeft: '20px !important',

      '& .MuiButton-label': {
        position: 'relative',
        top: '1px',
      },
    },
    borderPrimaryButton: {
      border: `2px solid ${STYLE.COLOR.ACCENT_PRIMARY}`,
      '& .MuiButton-label': {
        color: STYLE.COLOR.ACCENT_PRIMARY,
      },
      '&:hover': { border: `2px solid ${STYLE.COLOR.ACCENT_PRIMARY}` },
    },
    createButton: {
      fontSize: `${toREM(15)}`,
      lineHeight: '17px',
      textTransform: 'none',
      boxShadow: 'none',
      fontWeight: 600,
      float: 'right',
      outline: 'none',

      '& .MuiButton-label': {
        position: 'relative',
        top: '1px',
      },
    },
    primaryButtonPadding: {
      fontSize: `${toREM(STYLE.FONT.SIZE.PARAGRAPH)}`,
      border: `2px solid ${STYLE.COLOR.ACCENT_PRIMARY}`,
      paddingLeft: `${toREM(13)}`,
      paddingRight: `${toREM(13)}`,
      '& .MuiButton-label': {
        color: STYLE.COLOR.ACCENT_PRIMARY,
      },
    },
    primaryButtonLarge: {
      width: `${toREM(160)}`,
      height: `${toREM(50)}`,
    },
    primaryButtonDisableHover: {
      '&:hover': {
        background: STYLE.COLOR.ACCENT_PRIMARY,
      },
    },
    primaryButtonExtraLarge: {
      width: `${toREM(200)}`,
      height: `${toREM(50)}`,
    },
    primaryButtonLarger: {
      width: `${toREM(217)}`,
      height: `${toREM(50)}`,
    },
    editButton: {
      '&:focus': {
        outline: 'none',
        border: 'none',
      },

      '& .MuiButton-label:focus': {
        outline: 'none',
        border: 'none',
      },
    },
    drawerCTA: {
      width: '100%',
      textTransform: 'inherit',
      maxWidth: `${toREM(130)}`,
      height: `${toREM(35)}`,
      minWidth: `${toREM(120)}`,
      paddingTop: '6px',
      display: 'flex',
      alignItems: 'center',
      fontSize: `${toREM(STYLE.FONT.SIZE.PARAGRAPH)}`,
      lineHeight: 0,

      '&:hover': {
        backgroundColor: 'transparent',
      },
    },
    warningButton: {
      border: `2px solid ${STYLE.COLOR.ERROR_ICON}`,
      '& .MuiButton-label': {
        color: STYLE.COLOR.ERROR_ICON,
      },
    },
    deleteButton: {
      position: 'relative',
      top: '24px',
      paddingLeft: 0,
      paddingRight: 0,

      '&:hover': {
        backgroundColor: 'transparent',
      },
      '& svg': {
        '& path': {
          fill: STYLE.COLOR.ERROR,
        },
      },
    },
    hideDeleteButton: {
      display: 'none',
    },
    showDeleteButton: {
      display: 'flex',
    },
    addUserButton: {
      display: 'flex',
      justifyContent: 'flex-start',
      padding: 0,
      fontSize: `${toREM(STYLE.FONT.SIZE.PARAGRAPH)}`,
      textTransform: 'capitalize',
      width: `${toREM(128)}`,
      height: `${toREM(20)}`,
      position: 'relative',
      right: '2px',
      marginBottom: `${toREM(24)}`,
      fontWeight: 600,

      '&:hover': {
        backgroundColor: 'transparent',
      },

      '& .MuiButton-startIcon': {
        marginLeft: 0,
        marginRight: '10px',
      },

      '& svg': {
        '& path': {
          fill: STYLE.COLOR.SUCCESS,
        },
      },
    },
    noPadding: {
      padding: 0,
    },
    loadingPadding: {
      padding: 10,
    },
    noOutline: {
      outline: 'none',
    },
    nextButton: {
      height: `${toREM(36)}`,
      width: `${toREM(120)}`,
      fontSize: `${toREM(STYLE.FONT.SIZE.PARAGRAPH)}`,
      lineHeight: '17px',
      textTransform: 'none',
      boxShadow: 'none',
      fontWeight: 600,
      outline: 'none',

      '& .MuiButton-label': {
        position: 'relative',
        top: '1px',
      },
    },
    discardButton: {
      height: '36px',
      fontSize: `${toREM(STYLE.FONT.SIZE.PARAGRAPH)}`,
      lineHeight: '17px',
      textTransform: 'none',
      boxShadow: 'none',
      fontWeight: 600,
      marginRight: '17px',
    },
    floatingAppBarButton: {
      height: `${toREM(56)}`,
    },
    closeButtonWidth: {
      width: `${toREM(120)}`,
    },
    badgeButtonContainer: {
      height: 35,
      fontSize: `${toREM(STYLE.FONT.SIZE.PARAGRAPH)}`,
      fontWeight: 600,
      padding: '0 10px !important',

      '& .MuiButton-label': {
        position: 'relative',
        top: 1,
      },

      '& .MuiButton-endIcon': {
        '& svg': {
          opacity: 1,
          '& g': {
            '& polygon': {
              fill: STYLE.COLOR.WHITE,
            },
          },
        },
      },
    },
    badgeProjectActiveBadgeDeactivated: {
      backgroundColor: STYLE.COLOR.SECONDARY_DARKER,
      '&:hover': {
        backgroundColor: STYLE.COLOR.SECONDARY_DARKER,
      },
    },
    badgeActive: {
      backgroundColor: STYLE.COLOR.SUCCESS,
      '&:hover': {
        backgroundColor: STYLE.COLOR.SUCCESS,
      },
    },
    badgeRevoked: {
      backgroundColor: STYLE.COLOR.ERROR_ICON,
      '&:hover': {
        backgroundColor: STYLE.COLOR.ERROR_ICON,
      },
    },
    disabled: {
      opacity: 0.5,
    },
  })
);
