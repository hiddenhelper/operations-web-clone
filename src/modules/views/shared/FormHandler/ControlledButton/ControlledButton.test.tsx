import React from 'react';
import { render, fireEvent, act } from '@testing-library/react';
import { disableFocus, enableFocusOnTab } from '../../../../../utils/generalUtils';

jest.mock('../../../../../utils/generalUtils', () => {
  return {
    disableFocus: jest.fn(),
    enableFocusOnTab: jest.fn(),
  };
});

import ControlledButton from './ControlledButton';

describe('ControlledButton', () => {
  let wrapper;

  it('should render', () => {
    wrapper = render(
      <ControlledButton>
        <button>Testing</button>
      </ControlledButton>
    );
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it('should call disable and enable focus', () => {
    wrapper = render(
      <ControlledButton>
        <button>Testing</button>
      </ControlledButton>
    );

    act(() => {
      fireEvent.mouseDown(wrapper.getByTestId('controlled-button'));
      fireEvent.keyDown(wrapper.getByTestId('controlled-button'));
    });

    expect(disableFocus).toHaveBeenCalled();
    expect(enableFocusOnTab).toHaveBeenCalled();
  });
});
