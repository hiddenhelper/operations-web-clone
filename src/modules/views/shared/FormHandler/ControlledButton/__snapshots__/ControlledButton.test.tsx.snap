// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlledButton should render 1`] = `
<body>
  <div>
    <span
      class=""
      data-testid="controlled-button"
      tabindex="-1"
    >
      <button>
        Testing
      </button>
    </span>
  </div>
</body>
`;
